R/data.table.R

```diff
@@ -1881,17 +1881,64 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
 #    x
 #}
 
-
-as.matrix.data.table <- function(x,...)
-{
-  dm <- dim(x)
-  cn <- names(x)
+as.matrix.data.table <- function(x, rownames, ...) {
+  rn <- NULL
+  rnc <- NULL
+  if (!missing(rownames)) { # Convert rownames to a column index if possible
+    if (is.null(rownames)) {
+      warning("rownames is NULL, ignoring rownames")
+    } else if (length(rownames) == nrow(x)) {
+      # rownames argument is a vector of row names, no column in x to drop.
+      rn <- rownames
+      rnc <- NULL 
+    } else if (length(rownames) != 1) {
+      stop("rownames must be a single column in x or a vector of row names of length nrow(x)")
+    } else if (is.na(rownames)) {
+      warning("rownames is NA, ignoring rownames")
+    } else if (identical(rownames, FALSE)) {
+      warning("rownames is FALSE, ignoring rownames")
+    } else if (!(is.logical(rownames) || is.character(rownames) || is.numeric(rownames))) {
+      # E.g. because rownames is some sort of object that cant be converted to a column index
+      stop("rownames must be TRUE, a column index, a column name in x, or a vector of row names")
+    } else {
+      if (identical(rownames, TRUE)) {
+        if (haskey(x)) { 
+          rownames <- key(x) 
+          if (length(rownames) > 1) {
+            warning("rownames is TRUE but multiple keys found in key(x), using first column instead")
+            rownames <- 1
+          }
+        } else { 
+          rownames <- 1
+        }
+      }
+      if (is.character(rownames)) { # Handles cases where rownames is a column name, or key(x) from TRUE
+        rnc <- chmatch(rownames, names(x))
+        if (is.na(rnc)) stop(rownames, " is not a column of x")
+      } else { # rownames is an index already
+        if (rownames < 1 || rownames > ncol(x))
+          stop("rownames is ", rownames, " which is outside the column number range [1,ncol=", ncol(x), "]")
+        rnc <- rownames
+      }
+    }
+  }
+  # If the rownames argument has been used, and is a single column,
+  # extract that column's index (rnc) and drop it from x
+  if (!is.null(rnc)) { 
+    rn <- x[[rnc]]
+    dm <- dim(x) - c(0, 1)
+    cn <- names(x)[-rnc]
+    X <- x[, .SD, .SDcols = cn]
```

Comment:
`x <- x[, -..rnc]` now works here in dev instead of these 2 lines. A copy is needed in this case at this point, iiuc, otherwise, `x[, (rnc):=NULL]`  to remove that column by reference, currently.  And maybe `x[, ..rnc := NULL]` in future.

Reply:
Thanks, yes I discovered `-..` was implemented as I was working on this. Ultimately I decided not to use `-..rnc` as you then have to define a dummy `..rnc` at the top of the function to avoid `R CMD check --as-cran` complaining `..rnc` is a global variable with no visible binding.

Reply:
Interesting, I see.  That is a shame about the no visible binding note.
In general we've tried to use idiomatic data.table internally, and then deal with the no-visible-binding note explicitly by adding a dummy NULL as you say.  This way, when folk look at the internals they see how we use data.table ourselves.  If and when there is ever a solution for the CRAN note, we can fix it one distinct place by taking away the NULL definitions.